src/coprocessor/memoset/mod.rs

```diff
@@ -0,0 +1,832 @@
+        let kv = CircuitTranscript::make_kv(&mut cs.namespace(|| "kv"), g, s, key, value)?;
+        let new_transcript = transcript.add(&mut cs.namespace(|| "new_transcript"), g, s, &kv)?;
+
+        let acc_v = acc.hash();
+
+        let new_acc_v =
+            self.memoset
+                .synthesize_add(&mut cs.namespace(|| "new_acc_v"), acc_v, &kv)?;
+
+        let new_acc = AllocatedPtr::alloc_tag(
+            &mut cs.namespace(|| "new_acc"),
+            ExprTag::Num.to_field(),
+            new_acc_v,
+        )?;
+
+        Ok((new_acc, new_transcript.clone()))
+    }
+
+    fn synthesize_remove<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        g: &GlobalAllocator<F>,
+        s: &Store<F>,
+        acc: &AllocatedPtr<F>,
+        transcript: &CircuitTranscript<F>,
+        key: &AllocatedPtr<F>,
+        value: &AllocatedPtr<F>,
+    ) -> Result<(AllocatedPtr<F>, CircuitTranscript<F>), SynthesisError> {
+        let kv = CircuitTranscript::make_kv(&mut cs.namespace(|| "kv"), g, s, key, value)?;
+        let zptr = kv.get_value().unwrap_or(s.hash_ptr(&s.intern_nil())); // dummy case: use nil
+        let raw_count = self.memoset.count(&s.to_ptr(&zptr)) as u64; // dummy case: count is meaningless
+
+        let (kv_count, count) = CircuitTranscript::make_kv_count(
+            &mut cs.namespace(|| "kv_count"),
+            g,
+            s,
+            &kv,
+            raw_count,
+        )?;
+        let new_transcript = transcript.add(
+            &mut cs.namespace(|| "new_removal_transcript"),
+            g,
+            s,
+            &kv_count,
+        )?;
+
+        let new_acc_v = self.memoset.synthesize_remove_n(
+            &mut cs.namespace(|| "new_acc_v"),
+            acc.hash(),
+            &kv,
+            &count,
+        )?;
+
+        let new_acc = AllocatedPtr::alloc_tag(
+            &mut cs.namespace(|| "new_acc"),
+            ExprTag::Num.to_field(),
+            new_acc_v,
+        )?;
+        Ok((new_acc, new_transcript))
+    }
+
+    fn finalize<CS: ConstraintSystem<F>>(&mut self, cs: &mut CS, _g: &mut GlobalAllocator<F>) {
+        let r = self.memoset.allocated_r(cs);
+        enforce_equal(cs, || "r_matches_transcript", self.transcript.r(), &r);
+        enforce_equal_zero(cs, || "acc_is_zero", self.acc.clone().unwrap().hash());
+    }
+
+    fn synthesize_query<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        g: &GlobalAllocator<F>,
+        store: &Store<F>,
+        key: &AllocatedPtr<F>,
+        acc: &AllocatedPtr<F>,
+        transcript: &CircuitTranscript<F>,
+        not_dummy: &Boolean, // TODO: use this more deeply?
+    ) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>, CircuitTranscript<F>), SynthesisError> {
+        let value = key
+            .get_value()
+            .map(|k| {
+                self.queries_alloc
+                    .entry(k)
+                    .or_insert_with(|| {
```

Comment:
This may lead to non-deterministic circuit shapes, since the pattern of allocation may depend on the order of queries. In that case, we may need to allocate every time rather than reusing allocations as currently.

Reply:
Fixed in https://github.com/lurk-lab/lurk-rs/pull/1004/commits/efb747e358de898959199c070aab499a1a6c664a.